Improve attachment wide-layout display

squelart opened the pull request:
#### d0509c40b6a5d54a51c30b85c72cda34efda4077
<pre>
Improve attachment wide-layout display
<a href="https://bugs.webkit.org/show_bug.cgi?id=253687">https://bugs.webkit.org/show_bug.cgi?id=253687</a>
rdar://problem/106534843

Reviewed by Aditya Keerthi.

Some work towards the target vspecs:
- More complex shadow DOM tree that mirrors the specs.
- macOS rendering displays just an icon, with the new size.
- Better handling of long labels.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::attachmentPreviewAreaIdentifier):
(WebCore::attachmentInformationAreaIdentifier):
(WebCore::attachmentInformationBlockIdentifier):
(WebCore::attachmentSaveAreaIdentifier):
(WebCore::attachmentSaveIconIdentifier):
(WebCore::createContainedElement):
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
(WebCore::HTMLAttachmentElement::updateSaveButton):
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container):
(div#attachment-preview-area):
(attachment#attachment-preview):
(div#attachment-information-area):
(div#attachment-information-block):
(div#attachment-action):
(div#attachment-title):
(div#attachment-subtitle):
(div#attachment-save-area):
(button#attachment-save-button):
(div#attachment-save-icon):
* Source/WebCore/rendering/AttachmentLayout.mm:
(WebCore::AttachmentLayout::AttachmentLayout):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::paintAttachment):

Canonical link: <a href="https://commits.webkit.org/261665@main">https://commits.webkit.org/261665@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a0dbbfef99b98155cfeb65bc7060358d0bc5c36f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112307 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21445 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/940 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4086 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120923 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116371 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22783 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12737 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/4860 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118071 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100115 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105355 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/653 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45935 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13821 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/701 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/94308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14512 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19864 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52705 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8128 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16314 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/rendering/AttachmentLayout.mm:
```diff
@@ -163,6 +165,14 @@ static Color titleTextColorForAttachment(const RenderAttachment& attachment, Att
+        iconRect = FloatRect(0, 0, attachmentImageOnlyIconSize, attachmentImageOnlyIconSize);
+        iconBackgroundRect = iconRect;
+        attachmentRect = iconBackgroundRect;
+        attachmentRect = encloseRectToDevicePixels(attachmentRect, attachment.document().deviceScaleFactor());
```
```suggestion
        attachmentRect = encloseRectToDevicePixels(iconBackgroundRect, attachment.document().deviceScaleFactor());
```

pxlcoder commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -22,10 +22,13 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+
```
Unnecessary newline.

rr-codes commented on Source/WebCore/rendering/AttachmentLayout.mm:
```diff
@@ -47,6 +47,8 @@
 const CGFloat attachmentIconBackgroundRadius = 3;
 const CGFloat attachmentIconToTitleMargin = 2;
 
+const CGFloat attachmentImageOnlyIconSize = 52;
```
can these be constexpr?

rr-codes replied:
oh maybe because its CGFloat and not normal float it can't be? Not sure

pxlcoder replied:
`CGFloat` is just a typedef of `float` or `double` depending on the architecture.

squelart replied:
There's a mixture of `const`s and `constexpr`s, happy to change them all to the latter. 👍

cdumez commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -149,7 +179,7 @@ static QualifiedName saveAttr()
 
 template <typename ElementType>
-static Ref<ElementType> createContainedElement(HTMLElement& container, const AtomString& id, String textContent)
+static Ref<ElementType> createContainedElement(Element& container, const AtomString& id, String textContent = { })
```
`String&&` or `const String&`

squelart replied:
`String&&` produces identical code (at the cost of having to explicitly create a temporary in some callers -- that's why I initially preferred `String`), so I'll go with that.

Just curious: What is the reason to prefer passing by reference? Is this project-wide unspoken knowledge? It may be good to add it to https://webkit.org/code-style-guidelines/

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -234,20 +270,27 @@ class AttachmentSaveEventListener final : public EventListener {
-            m_containerElement->removeChild(*m_saveButton);
+            auto saveArea = m_saveButton->parentElement();
+            ASSERT(saveArea);
+            auto saveAreaContainer = saveArea->parentElement();
```
Can we store the container as a member, rather than rely on the DOM structure here?

Also, anywhere you assign the result of `parentElement()` to variable should probably use `RefPtr` rather than `auto` (raw pointer).

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -234,20 +270,27 @@ class AttachmentSaveEventListener final : public EventListener {
-    if (!m_saveButton) {
-        m_saveButton = createContainedElement<HTMLButtonElement>(*m_containerElement, attachmentSaveButtonIdentifier(), String { });
+    if (!m_saveButton && m_titleElement) {
+        auto informationBlock = m_titleElement->parentElement();
```
Same comment as above.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -234,20 +270,22 @@ class AttachmentSaveEventListener final : public EventListener {
-        return;
-
     if (!show) {
         if (m_saveButton) {
```
Any reason this can't just be `if (m_saveArea)`? Looks like `m_saveArea` and `m_saveButton` have the same lifetime?

Do we need both?

squelart replied:
Yes they have the same lifetimes.
But I believe we do need both: `m_saveArea` contains `m_saveButton` but they have different sizes; we need `m_saveButton` in code elsewhere to find its location in the attachment; and we need `m_saveArea` as the top element to be removed when we don't need the save button anymore (otherwise it would still take some space).

pxlcoder replied:
I expect we could achieve these requirements by using `display: flex` rather than `display: grid` in some places, but that would be a bigger change, and the current approach is also fine.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -234,20 +270,22 @@ class AttachmentSaveEventListener final : public EventListener {
 
-    if (!m_saveButton) {
-        m_saveButton = createContainedElement<HTMLButtonElement>(*m_containerElement, attachmentSaveButtonIdentifier(), String { });
+    if (!m_saveButton && m_titleElement) {
```
Wonder if this should also be `!m_saveArea`.

squelart replied:
Same-same, as they have the same lifetime. I prefer to mention the button here, as it's the most important element, the area is just a style-implementation detail.

squelart committed:
Improve attachment wide-layout display
https://bugs.webkit.org/show_bug.cgi?id=253687
rdar://problem/106534843

Reviewed by Aditya Keerthi.

Some work towards the target vspecs:
- More complex shadow DOM tree that mirrors the specs.
- macOS rendering displays just an icon, with the new size.
- Better handling of long labels.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::attachmentPreviewAreaIdentifier):
(WebCore::attachmentInformationAreaIdentifier):
(WebCore::attachmentInformationBlockIdentifier):
(WebCore::attachmentSaveAreaIdentifier):
(WebCore::attachmentSaveIconIdentifier):
(WebCore::createContainedElement):
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
(WebCore::HTMLAttachmentElement::updateSaveButton):
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container):
(div#attachment-preview-area):
(attachment#attachment-preview):
(div#attachment-information-area):
(div#attachment-information-block):
(div#attachment-action):
(div#attachment-title):
(div#attachment-subtitle):
(div#attachment-save-area):
(button#attachment-save-button):
(div#attachment-save-icon):
* Source/WebCore/rendering/AttachmentLayout.mm:
(WebCore::AttachmentLayout::AttachmentLayout):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::paintAttachment):

Canonical link: https://commits.webkit.org/261665@main
